libqtile/confreader.py

```diff
@@ -140,8 +140,17 @@ def validate(self) -> None:
         """
         Validate the configuration against the core.
         """
-        valid_keys = core.get_keys()
-        valid_mods = core.get_modifiers()
+        valid_keys = keysyms.keys()
+        valid_mods = [
+            "shift",
+            "lock",
+            "control",
+            "mod1",
+            "mod2",
+            "mod3",
+            "mod4",
+            "mod5",
+        ]
```

Comment:
I don't think we want to hard code this!

Reply:
Yeah, I agree. One thing to consider is that we validate the config in `qtile check`, so in theory we shouldn't need a running core to check. This works fine with the X backend (presumably why the functions were standalone functions in the core.py module). A possible approach might be to set `get_keys` and `get_modifiers` as abstract static methods on the base `Core` as part of the 'backend protocol'. The wayland backend should be able to ask xkbcommon for the data.

Reply:
> A possible approach might be to set `get_keys` and `get_modifiers` as abstract static methods on the base `Core` as part of the 'backend protocol'. The wayland backend should be able to ask xkbcommon for the data.

:+1:

Reply:
https://github.com/qtile/qtile/pull/4180 does this